Sources/DeepLearning/Layer.swift

```diff
@@ -1314,3 +1314,44 @@ public struct Reshape<Scalar: TensorFlowFloatingPoint>: Layer {
         return input.reshaped(toShape: shape)
     }
 }
+
+/// An input to a recurrent neural network.
+public struct RNNInput<TimeStepInput: Differentiable, State: Differentiable>: Differentiable {
+    /// The input at the current time step.
+    public var timeStepInput: TimeStepInput
+    /// The previous state.
+    public var previousState: State
+
+    @differentiable
+    public init(timeStepInput: StepInput, previousState: State) {
+        self.timeStepInput = timeStepInput
+        self.previousState = previousState
+    }
+}
+
+/// A recurrent neural network cell.
+public protocol RNNCell: Layer where Input == RNNInput<TimeStepInput, State> {
```

Comment:
I'm intentionally not adding an `init`. At this point, I have not reached a conclusion about whether `init` is generalizable across all different RNN cells.